src/tactic/core.lean

```diff
@@ -1164,21 +1180,32 @@ For example,
 adds an instance `ring new_int`, defined to be the instance of `ring ℤ` found by `apply_instance`.
 
 Multiple instances can be added with `@[derive [ring, module ℝ]]`.
+
+This derive handler applies only to declarations made using `def`, and will fail on such a
+declaration if it is unable to derive an instance. It is run with higher priority than the built-in
+handlers, which will fail on `def`s.
 -/
-@[derive_handler] meta def delta_instance : derive_handler :=
-λ cls tp,
-(do tp' ← mk_const tp,
-   tgt ← to_expr ``(%%cls %%tp'),
-   (_, v) ← solve_aux tgt (delta_target [tp] >> apply_instance >> done),
-   v ← instantiate_mvars v,
-   nm ← get_unused_name $ tp ++
-     match tgt with
-     | expr.app (expr.const nm _) _ := nm
+@[derive_handler, priority 2000] meta def delta_instance : derive_handler :=
+λ cls new_decl_name,
+do env ← get_env,
+if env.is_inductive new_decl_name then return ff else
+do new_decl_type ← declaration.type <$> get_decl new_decl_name,
+   new_decl_pexpr ← resolve_name new_decl_name,
+   tgt ← to_expr $ apply_under_pis cls new_decl_pexpr new_decl_type,
+   (_, inst) ← solve_aux tgt
+     (intros >> reset_instance_cache >> delta_target [new_decl_name]  >> apply_instance >> done),
+   inst ← instantiate_mvars inst,
+   tgt ← instantiate_mvars tgt,
+   nm ← get_unused_decl_name $ new_decl_name ++
+     match cls with
+     -- the postfix is needed because we can't protect this name. using nm.last directly can
+     -- conflict with open namespaces
+     | (expr.const nm _) := (nm.last ++ "_1" : string)
      | _ := "inst"
      end,
-   add_decl $ mk_definition nm [] tgt v,
+   add_decl $ mk_definition nm inst.collect_univ_params tgt inst,
    set_basic_attribute `instance nm tt,
-   return tt) <|> return ff
+   return tt
```

Comment:
What does this tactic now do when applied to a `def` but where `apply_instance` doesn't work? It seems like it fails. Does that prevent all future `derive_handlers` on `defs` with priority below 2000 to fire? That doesn't seem good to me.

Reply:
The working hypothesis (see the Zulip thread) is that this is the only derive handler we'll want to use for `def`s. If `@[derive...]` is applied to an inductive, this handler will return `ff` and the others will take over. If it's applied to a `def` and this fails, we'll see an error message from the type class resolution failure.

If the hypothesis is false, we'll need to change things. The derive handler API isn't really suited for classes with multiple handlers. We dance around that here by distinguishing `def`s and `inductive`s, but if that turns out not to work, we'll probably need a second attribute.

Reply:
That working hypothesis must be false. If `pi_instances` was a derive-handler, it would become dangerously close to a counterexample (it isn't, though). But it must be possible to write automation to construct instances on subtypes/quotient types (which are probably wrapped in a `def`). But we can work around that by not using the `derive` interface.

So the reason that we don't return `ff` but fail instead, is so that we get to see the error message?

Reply:
It's so that we see the specific error message from `delta_instance`. Otherwise you'll see it fail in two different ways. If it fails to derive a class without any built-in handlers (`@[derive ring]`) you'll get an error about no handlers being found. If it fails to derive e.g. `decidable_eq`, you'll get the error message from the dedicated `decidable_eq` handler, which will say something confusing about it not being an inductive.

Reply:
Ok, this seems fine then for now.